### DISCUSSION TITLE
integrating web sockets into nuxt3

### DISCUSSION THREAD

**Discussion by andreas83** [nuxt/nuxt#16701] [Questions]
**questions:** 
where is the right place to open a new websocket connection? 
plugins or composales?  

**bigger picture:**  
the app or to be more precise the user subscribes to various WS channels like, permission changed or content is locked by another user etc. 
 
**the trouble:**

everytime the composable or plugin is loaded a new WS object is created.

my ws.client.js plugin:

```
import { defineNuxtPlugin } from '#app'
export default defineNuxtPlugin(({ $config }) => {
  return {
    provide: {
      ws: () => new WebSocket($config.wsURL )
    }
  }
})

```
my useWebsockets.ts composable:
```
export const useWs = () => {
    if (!process.client) {
        return {};
    }
    const nuxtApp = useNuxtApp()

    let ws = new WebSocket(nuxtApp.$config.wsURL )

    
    return useState('websocket', () => ws)
}
```

**Comment by stafyniaksacha** [comment]
Hi @andreas83 !

I made a simple POC using ws with nitro node server. You can find it here: https://github.com/stafyniaksacha/nuxt-poc-websockets

It works via a custom nitro preset, so you do not have to run `node .output/server/index.mjs` but `node start.mjs` instead

For development, it's not possible to bind to nuxt server (which is served with a unix socket, proxied by vite) so it's why another server is created
